bundles/org.openhab.core.model.rule/src/org/openhab/core/model/rule/jvmmodel/RulesRefresher.java

```diff
@@ -29,51 +49,133 @@
+
+    private static final String POOL_NAME = "automation";
+    public static final String RULES_REFRESH = "rules_refresh";
 
     private final Logger logger = LoggerFactory.getLogger(RulesRefresher.class);
 
-    ModelRepository modelRepository;
-    private ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
     private ScheduledFuture<?> job;
+    private ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool(POOL_NAME);
+    private boolean started;
+    private ReadyMarker marker = new ReadyMarker("dsl", RULES_REFRESH);
 
-    public void setModelRepository(ModelRepository modelRepository) {
-        this.modelRepository = modelRepository;
-    }
+    private ItemRegistryChangeListener itemRegistryChangeListener;
+    private ThingRegistryChangeListener thingRegistryChangeListener;
+
+    @Reference
+    protected ModelRepository modelRepository;
+
+    @Reference
+    protected ItemRegistry itemRegistry;
+
+    @Reference
+    protected ThingRegistry thingRegistry;
 
-    public void unsetModelRepository(ModelRepository modelRepository) {
-        this.modelRepository = null;
+    @Reference
+    protected EventPublisher eventPublisher;
+
+    @Reference
+    protected ReadyService readyService;
+
+    @Activate
+    protected void activate() {
+        readyService.registerTracker(this, new ReadyMarkerFilter().withType("dsl").withIdentifier("rules"));
     }
 
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
     protected void addActionService(ActionService actionService) {
-        scheduleRuleRefresh();
+        if (started) {
+            scheduleRuleRefresh();
+        }
     }
 
     protected void removeActionService(ActionService actionService) {
-        scheduleRuleRefresh();
+        if (started) {
+            scheduleRuleRefresh();
+        }
     }
 
     protected synchronized void scheduleRuleRefresh() {
         if (job != null && !job.isDone()) {
             job.cancel(false);
         }
-        job = scheduler.schedule(runnable, REFRESH_DELAY, TimeUnit.MILLISECONDS);
-    }
-
-    Runnable runnable = new Runnable() {
-        @Override
-        public void run() {
+        job = scheduler.schedule(() -> {
             try {
-                if (modelRepository != null) {
-                    modelRepository.reloadAllModelsOfType("rules");
-                }
+                modelRepository.reloadAllModelsOfType("rules");
             } catch (Exception e) {
                 logger.debug("Exception occurred during execution: {}", e.getMessage(), e);
             }
+            readyService.markReady(marker);
+        }, REFRESH_DELAY, TimeUnit.MILLISECONDS);
+        readyService.unmarkReady(marker);
+    }
+
+    private void setStartLevel() {
+        if (!started) {
+            started = true;
+            // TODO: This is still a very dirty hack in the absence of a proper system start level management.
```

Comment:
Can this TODO be done in this PR?

Reply:
No, definitely not - this is about https://github.com/eclipse-archived/smarthome/issues/1896, which will be a huge new PR in itself.